lib/compat/wordpress-6.2/get-global-styles-and-settings.php

```diff
@@ -75,3 +75,122 @@ function _wp_theme_has_theme_json_clean_cache_upon_upgrading_active_theme( $upgr
 		}
 	}
 }
+
+/**
+ * Returns the stylesheet resulting of merging core, theme, and user data.
+ *
+ * @param array $types Types of styles to load. Optional.
+ *                     It accepts 'variables', 'styles', 'presets' as values.
+ *                     If empty, it'll load all for themes with theme.json support
+ *                     and only [ 'variables', 'presets' ] for themes without theme.json support.
+ *
+ * @return string Stylesheet.
+ */
+function gutenberg_get_global_stylesheet( $types = array() ) {
+	/**
+	 * Filters whether the cached global stylesheet can be used.
+	 *
+	 * @param boolean $can_use_cached Whether the cached global stylesheet can be used.
+	 */
+	$can_use_cached = apply_filters(
+		'wp_get_global_stylesheet_can_use_cache',
```

Comment:
What would be the use-case for having this controlled by a filter? If this is just to allow turning it off during development, I would suggest we remove this and rethink it, as this consideration does not only apply to the global stylesheet but also anything related to `theme.json`.

See my other comment - maybe a filter is actually the best approach, but then it should probably be a more general filter? Anyway, my main point is we should discuss this holistically, not across several PRs where we're adding different ideas in code. I think an issue to discuss first would be best for this.

At a minimum, we should align on one thing throughout those PRs, even if we want to remain open to changing it later. Right now I'm afraid the approaches are getting mixed up, which would be a problem.

Reply:
The use case for the filter is documented in this other conversation (it's unrelated to development/debug) https://github.com/WordPress/gutenberg/pull/45679#discussion_r1025470701

Reply:
@oandregal Thanks for the context. What I'm not understanding yet though, why is this specifically a problem here? Wouldn't the same concerns apply to pretty much any cache usage in WordPress? How is this one different?

Reply:
I think we should add this filter later, I am not sure about it either.

Reply:
> What would be the use-case for having this controlled by a filter?

One example is that this filter would allow plugins to preview styles in the frontend without saving them.

Plugins have the ability to change the styles using one of the `wp_theme_json_data_*` filters, but those filters are not applied if the global stylesheet is cached since `WP_Theme_JSON_Resolver_Gutenberg::get_merged_data()` is never executed.

Without a filter that short-circuits the cache, a plugin that hooks into `wp_theme_json_data_user` to show how styles would look like in the frontend won't be able to do so if the actual styles (the ones showed to visitors) are already cached.

The plugin could of course invalidate the cache before the styles are previewed, but then those previewed-not-actual-styles would be cached, and the plugin would need to invalidate the cache again before a visitor access the site.

With a filter, this would be much easier to handle.

Anyway, happy to remove the filter for now and leave it for a follow-up.

Reply:
I may have been too terse on the rationale, let me rephrase/provide more context.

This is the product need we want to support: 3rd parties are able to hook into the [theme.json data filters](https://make.wordpress.org/core/2022/10/10/filters-for-theme-json-data/), which control the output of the `gutenberg_get_global_stylesheet` method. Those filters only run if the value is recalculated, not taken from the cache. There are valid uses cases that are not covered by the current cache invalidation, hence we need a mechanism for 3rd parties to force the values to be recalculated. Some of these use cases are:

- a hosting company may want to recalculate the stylesheet based on a user profile.
- an agency may want to control the styles of a site based on seasonal events (e.g.: black friday).
- a plugin may want to pre-visualize styles in the front-end that were not saved.

That's the why, the product need. I hope this provides a clear context. 

Now, discussing implementations to support this product need:

- I pushed against the filter initially ([see](https://github.com/WordPress/gutenberg/pull/41201#issuecomment-1303111763)). I though asking consumers to clean the cache would be enough.
- Upon further inspection, I realized that the code was vulnerable to race conditions ([see](https://github.com/WordPress/gutenberg/pull/45679#discussion_r1025470701)), an issue that the filter approach does not have.

Is there another alternative we should consider to be able to support this use case?

Reply:
I've talked about this with Felix in private. I understand Felix's point of view, though I have a different perspective. In the interest of making progress, let's leave the filter out of this PR and re-evaluate later, should we need.

Reply:
Thanks for the feedback @felixarntz. I see now how requiring plugin developers to disable the cache to be able to hook into the `wp_theme_json_data_*` filters shouldn't be a required pattern, so I removed the filter for now in https://github.com/WordPress/gutenberg/pull/45679/commits/2b163269b77617a597ce2b5e69c568f3742d3844.